tempesta_fw/classifier/frang.c

```diff
@@ -0,0 +1,905 @@
+	}
+
+	/*
+	 * Increment connection counters ever if we return TFW_BLOCK.
+	 * Synchronous sockets will call connection_drop callback,
+	 * so our frang_conn_close() is also called and we decrement
+	 * conn_curr there, but leave conn_new as is - we account failed
+	 * connection tries as well as successfully establised connections.
+	 */
+	ra->history[i].conn_new++;
+	ra->conn_curr++;
+
+	if (frang_cfg.conn_max && ra->conn_curr > frang_cfg.conn_max)
+		return TFW_BLOCK;
+	if (frang_cfg.req_burst && ra->history[i].req > frang_cfg.req_burst)
+		return TFW_BLOCK;
+
+	/* Collect new connections sum. */
+	for (i = 0; i < FRANG_FREQ; i++)
+		if (ra->history[i].ts + FRANG_FREQ >= ts)
+			csum += ra->history[i].conn_new;
+	if (frang_cfg.conn_rate && csum > frang_cfg.conn_rate)
+		return TFW_BLOCK;
+
+	return TFW_PASS;
+}
+
+static int
+frang_conn_new(struct sock *sk)
+{
+	return frang_account_do(sk, frang_conn_limit);
+}
+
+static int
+__frang_conn_close(FrangAcc *ra, struct sock *unused)
+{
+	BUG_ON(!ra->conn_curr);
+
+	ra->conn_curr--;
+
+	return TFW_PASS;
+}
+
+/**
+ * Just update current connection count for the user.
+ */
+static int
+frang_conn_close(struct sock *sk)
+{
+	return frang_account_do(sk, __frang_conn_close);
+}
+
+static int
+frang_req_limit(FrangAcc *ra, struct sock *sk)
+{
+	unsigned long ts = jiffies * FRANG_FREQ / HZ;
+	unsigned int rsum = 0;
+	int i = ts % FRANG_FREQ;
+
+	if (ra->history[i].ts != ts) {
+		ra->history[i].ts = ts;
+		ra->history[i].conn_new = 0;
+		ra->history[i].req = 0;
+	}
+	ra->history[i].req++;
+
+	if (frang_cfg.req_burst && ra->history[i].req > frang_cfg.req_burst)
+		goto block;
+
+	/* Collect current request sum. */
+	for (i = 0; i < FRANG_FREQ; i++)
+		if (ra->history[i].ts + FRANG_FREQ >= ts)
+			rsum += ra->history[i].req;
+	if (frang_cfg.req_rate && rsum > frang_cfg.req_rate)
+		goto block;
+
+	return TFW_PASS;
+
+block:
+	/*
+	 * TODO reset connection instead of wasting resources
+	 * on gentle closing. See ss_do_close() in sync_socket.
+	 */
+	TFW_DBG("%s: close connection\n", __FUNCTION__);
+	ss_close(sk);
+	return TFW_BLOCK;
+}
+
+static int
+frang_http_uri_len(const TfwHttpReq *req)
+{
+	/* FIXME: tfw_str_len() iterates over chunks to calculate the length.
+	 * This is too slow. The value must be stored in a TfwStr field. */
+	if (tfw_str_len(&req->uri_path) > frang_cfg.http_uri_len) {
+		TFW_DBG("frang: http_uri_len limit reached\n");
+		return TFW_BLOCK;
+	}
+	return TFW_PASS;
+}
+
+static int
+frang_http_field_len_raw(const TfwHttpReq *req)
+{
+	const TfwStr *field, *end;
+
+	FOR_EACH_HDR_FIELD_FROM(field, end, req, req->hdr_rawid) {
+		if (tfw_str_len(field) > frang_cfg.http_field_len) {
+			TFW_DBG("frang: http_field_len limit reached\n");
+			return TFW_BLOCK;
+		}
+	}
+	return TFW_PASS;
+}
+
+static int
+frang_http_field_len_special(const TfwHttpReq *req)
+{
+	const TfwStr *field, *end;
+
+	FOR_EACH_HDR_FIELD_SPECIAL(field, end, req) {
+		if (tfw_str_len(field) > frang_cfg.http_field_len) {
```

Comment:
This function and the above check a header length against the same configuration argument. However, the task was check header length for Cookie (not all "special" headers) and all other headers separately since Cookie is usually much bigger than any other headers.

So (1) there should be different configuration options and (2) we're interested in Cookie header and 'all other' (special and raw) headers.